web/packages/teleterm/src/ui/services/modals/modalsService.ts

```diff
@@ -58,20 +67,44 @@ export class ModalsService extends ImmutableStore<State> {
    * old dialog with the new one.
    * The old dialog is canceled, if possible.
    *
-   * The returned closeDialog function can be used to close the dialog and automatically call the
-   * dialog's onCancel callback (if present).
+   * The passed `abortSignal` or the returned `closeDialog` function can be used to close the dialog
+   * and automatically call the dialog's onCancel callback (if present).
    */
-  openRegularDialog(dialog: Dialog): { closeDialog: () => void } {
+  openRegularDialog(
+    dialog: Dialog,
+    abortSignal?: AbortSignal
+  ): {
+    closeDialog: () => void;
+  } {
+    const onCancelDialog = () => dialog['onCancel']?.();
+    const sharedSignal = this.getSharedAbortSignal(abortSignal);
+    if (sharedSignal.aborted) {
+      onCancelDialog();
+      return {
+        closeDialog: () => {},
+      };
+    }
+
     this.state.regular?.['onCancel']?.();
+
+    const closeDialog = () => {
+      sharedSignal.removeEventListener('abort', closeDialog);
```

Comment:
`sharedSignal.addEventListener('abort', closeDialog, {once: true})` instead of removing it here?

Reply:
Ah, nevermind, we have to handle the returned `closeDialog` being called by hand and not through the signal. We should have a test case for this, as in call `closeDialog` by hand, then abort the signal, then verify that the callback was called only once.